ops/model.py

```diff
@@ -1348,6 +1349,73 @@ class InvalidStatusError(ModelError):
     """Raised if trying to set an Application or Unit status to something invalid."""
 
 
+def _validate_action_result_key(key: str) -> None:
+    """Use regular expressions to validate that the given key is in the correct format.
+
+    Correct format is defined as starting with and ending with a lowercase alphanumeric
+    character, and containing nothing but lowercase, alphanumeric characters, hyphens and
+    periods.
+
+    Arguments:
+        key: a string representing the key for an item in an action result.
+
+    Raises:
+        ValueError: if a dict is passes with a key that fails to meet the format requirements.
+    """
+    # Existing charms may well be passing dicts to set_result that contain dicts pre-formatted
+    # to use dotted notation having worked around this in the past. We will deprecate this in the
+    # future to avoid possible duplication of keys.
+    # TODO: Drop this in a future release
+    if '.' in key:
+        msg = ("a '.' was detected in the dict key: '{}' while formatting a dict passed to "
+               "ActionEvent.set_result(). Keys using the dotted notation will be rejected in a "
+               "later release.")
+        warnings.warn(msg.format(key), DeprecationWarning)
+
+    # TODO: Drop the period from this regex when the above DeprecationWarning is removed
+    key_pattern = re.compile(r'^[a-z0-9](([a-z0-9-.]+)?[a-z0-9])?$')
+
+    if not key_pattern.match(key):
+        raise ValueError("Key '{}' is invalid: keys must start and end with lowercase "
+                         "alphanumeric, and contain only lowercase alphanumeric hyphens and "
+                         "periods".format(key))
+
+
+def _format_action_result_dict(dictionary: dict, parent_key: str = None) -> dict:
+    """Turn a nested dictionary into a flattened dictionary, using '.' as a key seperator.
+
+    This is used to allow nested dictionaries to be translated into the dotted format required by
+    the Juju `action-set` hook tool in order to set nested data on an action.
+
+    Additionally, this method performs some validation on keys to ensure they only use permitted
+    characters.
+
+    Example::
+
+        test_dict = {'a': {'b': 1, 'c': 2}}
+        flat_dict = _format_action_result_dict(test_dict)
+        # flat_dict is now {'a.b': 1, 'a.c': 2}
```

Comment:
Not a big deal, but what about turning this example into a runnable doctest with `>>>`? Something like this:

```
>>> test_dict = {'a': {'b': 1, 'c': 2}}
>>> _format_action_result_dict(test_dict)
{'a.b': 1, 'a.c': 2}
```

It's possible it wouldn't work conssitently on Python 3.5 due to dictionaries not being ordered on Python <3.6, but I'm not sure.

Reply:
Nice, will add in next commit!